### PR TITLE
Onboard Action wise system authz policy

### PR DESCRIPTION
**Pull request by senthalan**
### Purpose
This pull request introduces a new hierarchy resolver for Organization Units (OUs), enabling the authorization service to traverse OU trees without causing recursive authorization checks or import cycles. The OU initialization now returns this resolver for injection into the authorization service, and all relevant tests and mocks are updated to accommodate this change. Additionally, the `ResourceType` enum is extended to include user schema resources.

Part of https://github.com/asgardeo/thunder/pull/1500

Core integration of OU hierarchy resolver:

* Added the `ouHierarchyAdapter` implementation in `hierarchy_resolver.go`, providing direct access to OU hierarchy traversal methods (`IsAncestorOrSelf`, `GetAncestorOUIDs`) for use by the authorization service, bypassing authorization checks to prevent recursion.
* Modified `init.go` so that `Initialize` returns the OU service, hierarchy resolver, and exporter, enabling two-phase initialization and injection of the resolver into the authorization service.
* Updated `servicemanager.go` to inject the hierarchy resolver into the authorization service after OU initialization, breaking the import cycle between sysauthz and ou.

Testing and mocking updates:

* Added `hierarchy_resolver_test.go` with comprehensive unit tests for the new resolver, covering ancestor checks and ancestor listing across various edge cases.
* Updated `init_test.go` to validate the presence and correctness of the hierarchy resolver in all initialization scenarios. [[1]](diffhunk://#diff-5455e2bc566ee94c1f3a6251f007613eeea9bebd35833c0c7fbe4d3d08d7c2ffL64-R69) [[2]](diffhunk://#diff-5455e2bc566ee94c1f3a6251f007613eeea9bebd35833c0c7fbe4d3d08d7c2ffL84-R90) [[3]](diffhunk://#diff-5455e2bc566ee94c1f3a6251f007613eeea9bebd35833c0c7fbe4d3d08d7c2ffL104-R111) [[4]](diffhunk://#diff-5455e2bc566ee94c1f3a6251f007613eeea9bebd35833c0c7fbe4d3d08d7c2ffL124-R132) [[5]](diffhunk://#diff-5455e2bc566ee94c1f3a6251f007613eeea9bebd35833c0c7fbe4d3d08d7c2ffL140-R149) [[6]](diffhunk://#diff-5455e2bc566ee94c1f3a6251f007613eeea9bebd35833c0c7fbe4d3d08d7c2ffL160-R170) [[7]](diffhunk://#diff-5455e2bc566ee94c1f3a6251f007613eeea9bebd35833c0c7fbe4d3d08d7c2ffL187-R197) [[8]](diffhunk://#diff-5455e2bc566ee94c1f3a6251f007613eeea9bebd35833c0c7fbe4d3d08d7c2ffL203-R220)
* Extended the system authorization service mock to support the new `SetOUHierarchyResolver` method for testing.

Permissions and resource types:

* Added `ResourceTypeUserSchema` to the `ResourceType` enum in `permissions.go`, expanding supported resource categories.

### Related Issues
- N/A

### Related PRs
- N/A

### Checklist
- [ ] Followed the contribution guidelines.
- [ ] Manual test round performed and verified.
- [ ] Documentation provided. (Add links if there are any)
    - [ ] Ran Vale and fixed all errors and warnings
- [ ] Tests provided. (Add links if there are any)
    - [ ] Unit Tests
    - [ ] Integration Tests
- [ ] Breaking changes. (Fill if applicable)
    - [ ] Breaking changes section filled.
    - [ ] `breaking change` label added.

### Security checks
- [ ] Followed secure coding standards in [WSO2 Secure Coding Guidelines](https://security.docs.wso2.com/en/latest/security-guidelines/secure-engineering-guidelines/secure-coding-guidlines/introduction/)
- [ ] Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.

## Summary by CodeRabbit

* **New Features**
  * Added user schema as a manageable resource with Create/Read/Update/Delete/List permissions.
  * Enabled OU-hierarchy-aware resolver so parent OUs can grant read/list visibility to descendant resources.

* **Improvements**
  * Authorization now supports per-action OU inheritance for read access with clearer policy resolution and safer fallback behavior.

### CODE REVIEW

File: backend/internal/system/security/permissions.go [hunk 1]
```diff
@@ -149,6 +164,13 @@ var actionPermissionMap = map[Action]string{
...
 	ActionDeleteGroup: PermissionGroup,
 	ActionListGroups:  PermissionGroupView,
+
+	// User schema actions.
+	ActionCreateUserSchema: PermissionUserSchema,
+	ActionReadUserSchema:   PermissionUserSchemaView,
+	ActionUpdateUserSchema: PermissionUserSchema,
+	ActionDeleteUserSchema: PermissionUserSchema,
+	ActionListUserSchemas:  PermissionUserSchemaView,
 }
```
[the remark this answers is not in the document]
Comment (senthalan) [reply backend/internal/system/security/permissions.go:174]: This will be incorporated with a next PR

File: backend/internal/ou/hierarchy_resolver.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+					log.String("ouID", *current))
+				return false, nil
+			}
+			logger.Error("Failed to traverse organization unit hierarchy during ancestry check",
+				log.Error(err))
+			return false, &ErrorInternalServerError
+		}
+
+		current = ou.Parent
+	}
```
[the remark this answers is not in the document]
Comment (senthalan) [reply backend/internal/ou/hierarchy_resolver.go]: Can you analyse the OU implementation and see whether do we have the cyclic parent reference possible ?

File: backend/internal/ou/hierarchy_resolver.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+	return false, nil
+}
+
+// GetAncestorOUIDs returns ouID itself followed by every ancestor OU ID, walking up to
+// the root of the tree. The returned slice always contains at least ouID itself.
+//
+// If the chain is broken (an OU in the hierarchy is not found), the walk stops and a
+// ServiceError is returned so callers can handle incomplete results explicitly.
+func (r *ouHierarchyAdapter) GetAncestorOUIDs(
```
Comment (rajithacharith) [inline backend/internal/ou/hierarchy_resolver.go:101]: Is there a reason to have the ouId itself?

File: backend/internal/ou/hierarchy_resolver.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	ctx context.Context, ouID string,
+) ([]string, *serviceerror.ServiceError) {
+	logger := log.GetLogger().With(log.String(log.LoggerKeyComponentName, loggerComponentNameHierarchyResolver))
+
+	if ouID == "" {
+		return []string{}, nil
+	}
+
+	var result []string
+	current := &ouID
```
Comment (rajithacharith) [inline backend/internal/ou/hierarchy_resolver.go]: Do we need to use the pointer

File: backend/internal/ou/hierarchy_resolver.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+func (r *ouHierarchyAdapter) IsAncestorOrSelf(
+	ctx context.Context, ancestorOUID, descendantOUID string,
+) (bool, *serviceerror.ServiceError) {
+	logger := log.GetLogger().With(log.String(log.LoggerKeyComponentName, loggerComponentNameHierarchyResolver))
+
+	if ancestorOUID == "" || descendantOUID == "" {
+		return false, nil
+	}
+
+	current := &descendantOUID
```
Comment (rajithacharith) [inline backend/internal/ou/hierarchy_resolver.go]: Do we need to use pointers?

File: backend/internal/system/sysauthz/service.go [hunk 6]
```diff
@@ -44,18 +44,49 @@ type SystemAuthorizationServiceInterface interface {
...
+		},
+	}
+}
+
+// SetOUHierarchyResolver injects the OU hierarchy resolver into the service.
+// It is called once at application startup after the ou package is initialized.
+// The ouInheritancePolicy is built once here and reused for every subsequent authz call.
+func (s *systemAuthorizationService) SetOUHierarchyResolver(resolver OUHierarchyResolver) {
+	if resolver == nil {
+		s.policies.inheritancePolicy = nil
```
Comment (rajithacharith) [inline backend/internal/system/sysauthz/service.go]: Do we need this unset (sort of) capability

File: backend/internal/ou/hierarchy_resolver.go (hunk 3 above)
Comment (senthalan) [reply backend/internal/ou/hierarchy_resolver.go:101]: There needs to be some more changes needed to address this. Can I do it with another PR?
